### ISSUE TITLE
Unnecessary dom create and remove.

### ISSUE THREAD

**Issue by windyGex** [developit/preact#932]
```jsx
class A extends Component {
    render() {
        return <div>123</div>
    }
    componentWillUnmount() {
        console.log('unmount')
    }
    componentDidMount() {
        console.log('didMount')
    }
    componentDidUpdate() {
        console.log('didUpdate')
    }
}

class Test extends Component {

    render() {
        const visible = this.props.visible
        return (<div>
            {visible ? <div></div> : null}
            <div>
                {visible ? <div></div> : null}
                {this.props.children}
            </div>
        </div>);
    }
}

class App extends Component {
    state = { visible: false }

    setVisible(visible) {
        this.setState({
            visible
        });
    }
    render() {
        return (<div>
            <Test visible={this.state.visible}>
                <A />
            </Test>
            <button onClick={this.setVisible.bind(this, true)} type="primary">Commit</button>
            <button onClick={this.setVisible.bind(this, false)}>Cancel</button>
        </div>);
    }
}
```
When App visible switched,  A component always log `unmount` & `didMount`, won't log `didUpdate`. I think it should be effects using dom diff with vnode, actually, A component should always log `didUpdate`, because vnode that App rendered  never changed.

@developit

**Comment by yaodingyd** [comment]
This is because in [diff](https://github.com/developit/preact/blob/master/src/vdom/diff.js#L206-L216), preact wants to pluck a node of the same type from the existing children for unkeyed node to decrease diff work, and it would lead to issues when siblings have same node type, and in this case, it actually cause more diff work.

I believe the intention here is when adding/removing siblings, we still find the original node instead of the newly added/removed node to do diffing, but only using "same node type" seems like a over simplification.

**Comment by windyGex** [comment]
Well, If A component is a very complex composite component,  it cause some other effects。 for example，make multiple requests(didMount)，unexpected abort for request(unmount), wasted render performance.

**Comment by yaodingyd** [comment]
Duplicate of #857.

**Comment by marvinhagemeister** [comment]
This is fixed with #1440 🎉💯
